mypy fails Python 3.13

```sh
src/xtgeo/grid3d/_roff_parameter.py:164: error: "ndarray[tuple[Any, ...], dtype[signedinteger[_32Bit]]]" has no attribute "filled"  [attr-defined]
src/xtgeo/grid3d/_roff_parameter.py:169: error: "ndarray[tuple[Any, ...], dtype[floating[_32Bit]]]" has no attribute "filled"  [attr-defined]
src/xtgeo/grid3d/_gridprop_roxapi.py:232: error: Incompatible types in assignment (expression has type "ndarray[tuple[Any, ...], dtype[generic[Any]]]", variable has type "MaskedArray[tuple[Any, ...], dtype[Any]]")  [assignment]
src/xtgeo/grid3d/_gridprop_roxapi.py:232: error: Argument "casting" to "astype" of "ndarray" has incompatible type "Literal['no', 'equiv', 'safe', 'same_kind', 'unsafe'] | None"; expected "Literal['no', 'equiv', 'safe', 'same_kind', 'same_value', 'unsafe']"  [arg-type]
src/xtgeo/grid3d/_gridprop_lowlevel.py:82: error: Incompatible default for argument "dtype" (default has type "None", argument has type "type[Any] | dtype[Any] | _HasDType[dtype[Any]] | _HasNumPyDType[dtype[Any]] | tuple[Any, Any] | list[Any] | _DTypeDict | str")  [assignment]
src/xtgeo/grid3d/_gridprop_lowlevel.py:82: note: PEP 484 prohibits implicit Optional. Accordingly, mypy has changed its default to no_implicit_optional=True
src/xtgeo/grid3d/_gridprop_lowlevel.py:82: note: Use https://github.com/hauntsaninja/no_implicit_optional to automatically upgrade your codebase
src/xtgeo/grid3d/_gridprop_lowlevel.py:107: error: Incompatible types in assignment (expression has type "ndarray[tuple[Any, ...], dtype[signedinteger[_32Bit]]] | ndarray[tuple[Any, ...], dtype[Any]]", variable has type "MaskedArray[tuple[Any, ...], dtype[Any]]")  [assignment]
src/xtgeo/grid3d/_gridprop_lowlevel.py:109: error: Incompatible types in assignment (expression has type "ndarray[tuple[Any, ...], dtype[Any]]", variable has type "MaskedArray[tuple[Any, ...], dtype[Any]]")  [assignment]
src/xtgeo/grid3d/_gridprop_lowlevel.py:111: error: Incompatible types in assignment (expression has type "ndarray[tuple[Any, ...], dtype[Any]]", variable has type "MaskedArray[tuple[Any, ...], dtype[Any]]")  [assignment]
src/xtgeo/grid3d/_gridprop_lowlevel.py:112: error: Incompatible types in assignment (expression has type "ndarray[tuple[Any, ...], dtype[Any]]", variable has type "MaskedArray[tuple[Any, ...], dtype[Any]]")  [assignment]
src/xtgeo/grid3d/_gridprop_lowlevel.py:115: error: Incompatible types in assignment (expression has type "ndarray[tuple[Any, ...], dtype[Any]]", variable has type "MaskedArray[tuple[Any, ...], dtype[Any]]")  [assignment]
src/xtgeo/grid3d/grid_property.py:721: error: Need type annotation for "values"  [var-annotated]
src/xtgeo/grid3d/grid_property.py:744: error: Incompatible types in assignment (expression has type "ndarray[tuple[Any, ...], dtype[float64]]", variable has type "MaskedArray[tuple[Any, ...], dtype[Any]]")  [assignment]
src/xtgeo/grid3d/grid_property.py:747: error: Incompatible types in assignment (expression has type "ndarray[tuple[Any, ...], dtype[signedinteger[_32Bit]]]", variable has type "MaskedArray[tuple[Any, ...], dtype[Any]]")  [assignment]
src/xtgeo/grid3d/grid_property.py:1053: error: Incompatible types in assignment (expression has type "ndarray[tuple[Any, ...], dtype[Any]]", variable has type "MaskedArray[tuple[Any, ...], dtype[Any]]")  [assignment]
src/xtgeo/grid3d/grid_property.py:1056: error: Incompatible return value type (got "ndarray[tuple[int], dtype[Any]]", expected "MaskedArray[tuple[Any, ...], dtype[Any]]")  [return-value]
src/xtgeo/grid3d/grid_property.py:1058: error: Incompatible return value type (got "ndarray[tuple[Any, ...], dtype[Any]]", expected "MaskedArray[tuple[Any, ...], dtype[Any]]")  [return-value]
src/xtgeo/grid3d/grid_property.py:1058: error: Argument 1 to "filled" of "MaskedArray" has incompatible type "Buffer | _SupportsArray[dtype[Any]] | _NestedSequence[_SupportsArray[dtype[Any]]] | complex | bytes | str | _NestedSequence[complex | bytes | str]"; expected "complex | str | bytes | generic[Any] | None"  [arg-type]
Found 17 errors in 4 files (checked 118 source files)
```

--- a/.github/workflows/mypy.yml
+++ b/.github/workflows/mypy.yml
@@ -20,7 +20,7 @@ jobs:
   mypy:
     strategy:
       matrix:
-        python-version: ["3.10", "3.13"]
+        python-version: ["3.11", "3.13"]
     runs-on: ubuntu-latest
     steps:
       - uses: actions/checkout@v6

--- a/src/xtgeo/grid3d/_gridprop_lowlevel.py
+++ b/src/xtgeo/grid3d/_gridprop_lowlevel.py
@@ -2,7 +2,7 @@
 
 from __future__ import annotations
 
-from typing import TYPE_CHECKING, Literal
+from typing import TYPE_CHECKING, Any, Literal
 
 import numpy as np
 import numpy.ma as ma
@@ -79,7 +79,7 @@ def update_carray(
     self: GridProperty,
     undef: int | float | None = None,
     discrete: bool | None = None,
-    dtype: DTypeLike = None,
+    dtype: DTypeLike | None = None,
     order: Literal["C", "F", "A", "K"] = "F",
 ) -> cArray:
     """Copy (update) values from numpy to SWIG, 1D array, returns a pointer
@@ -101,20 +101,21 @@ def update_carray(
 
     logger.debug("Entering conversion from numpy to C array ...")
 
-    values = self.values.copy()
+    values_masked: ma.MaskedArray[Any] = ma.array(self.values, copy=True)
 
-    if not dtype:
-        values = values.astype(np.int32) if dstatus else values.astype(np.float64)
+    if dtype is None:
+        target_dtype: DTypeLike = np.int32 if dstatus else np.float64
+        values_masked = ma.array(values_masked.astype(target_dtype), copy=False)
     else:
-        values = values.astype(dtype)
+        values_masked = ma.array(values_masked.astype(dtype), copy=False)
 
-    values = ma.filled(values, undef)
-    values = np.asfortranarray(values)
+    values_filled = ma.filled(values_masked, undef)
+    values_filled = np.asfortranarray(values_filled)
 
     if order == "F":
-        values = np.asfortranarray(values)
+        values_filled = np.asfortranarray(values_filled)
 
-    values1d = np.ravel(values, order=order)
+    values1d = np.ravel(values_filled, order=order)
 
     if values1d.dtype == "float64" and dstatus and not dtype:
         values1d = values1d.astype("int32")

--- a/src/xtgeo/grid3d/_gridprop_roxapi.py
+++ b/src/xtgeo/grid3d/_gridprop_roxapi.py
@@ -118,7 +118,8 @@ def export_prop_roxapi(
     gname: str,
     pname: str,
     realisation: int = 0,
-    casting: Literal["no", "equiv", "safe", "same_kind", "unsafe"] | None = "unsafe",
+    casting: Literal["no", "equiv", "safe", "same_kind", "same_value", "unsafe"]
+    | None = "unsafe",
 ) -> None:
     """Export (i.e. store or save) to a Property icon in RMS via ROXAR API spec."""
     rox = RoxUtils(project, readonly=False)
@@ -192,7 +193,7 @@ def _store_in_roxar(
     pname: str,
     roxgrid: RoxarGrid3DType,
     realisation: int,
-    casting: Literal["no", "equiv", "safe", "same_kind", "unsafe"] | None,
+    casting: Literal["no", "equiv", "safe", "same_kind", "same_value", "unsafe"] | None,
 ) -> None:
     """Store property in RMS."""
     indexer = roxgrid.get_grid(realisation=realisation).grid_indexer
@@ -206,7 +207,7 @@ def _store_in_roxar(
         else roxtype.GridPropertyType.continuous
     )
 
-    val3d = self.values.copy()
+    val3d: np.ma.MaskedArray = np.ma.array(self.values, copy=True)
 
     dtype = self._roxar_dtype
     logger.info("DTYPE is %s for %s", dtype, pname)
@@ -229,7 +230,9 @@ def _store_in_roxar(
     _validate_dtype_in_roxar(val3d, original_dtype, dtype, casting == "unsafe")
 
     dtype = original_dtype
-    val3d = val3d.astype(dtype, casting=casting)
+    cast_mode: Literal["no", "equiv", "safe", "same_kind", "same_value", "unsafe"]
+    cast_mode = casting if casting is not None else "unsafe"
+    val3d = np.ma.array(val3d.astype(dtype, casting=cast_mode), copy=False)
 
     cellno = indexer.get_cell_numbers_in_range((0, 0, 0), indexer.dimensions)
 

--- a/src/xtgeo/grid3d/_roff_parameter.py
+++ b/src/xtgeo/grid3d/_roff_parameter.py
@@ -153,27 +153,29 @@ class RoffParameter:
             code_values = np.array(list(gridproperty.codes.keys()), dtype=np.int32)
 
         if not np.ma.isMaskedArray(gridproperty.values):
-            values = np.ma.masked_greater(
+            values: np.ma.MaskedArray = np.ma.masked_greater(
                 gridproperty.values,
                 UNDEF_INT_LIMIT if gridproperty.isdiscrete else UNDEF_LIMIT,
             )
         else:
-            values = gridproperty.values
+            values = np.ma.array(gridproperty.values, copy=False)
 
         if gridproperty.isdiscrete:
-            values = values.astype(np.int32).filled(-999)
+            values = np.ma.array(values.astype(np.int32), copy=False)
+            filled_values = values.filled(-999)
         else:
             # Although the roff format can contain double,
             # double typed parameters are not read by RMS so we
             # need to convert to float32 here
-            values = values.astype(np.float32).filled(-999.0)
+            values = np.ma.array(values.astype(np.float32), copy=False)
+            filled_values = values.filled(-999.0)
 
         return RoffParameter(
             nx=gridproperty.ncol,
             ny=gridproperty.nrow,
             nz=gridproperty.nlay,
             name=gridproperty.name or "",
-            values=np.asarray(np.flip(values, -1).ravel()),
+            values=np.asarray(np.flip(filled_values, -1).ravel()),
             code_names=code_names,
             code_values=code_values,
         )

--- a/src/xtgeo/grid3d/grid_property.py
+++ b/src/xtgeo/grid3d/grid_property.py
@@ -711,6 +711,7 @@ class GridProperty(_Grid3D):
             else None
         )
 
+        values: np.ma.MaskedArray
         if isinstance(invalues, (int, float)):
             vals = np.ma.zeros((ncol, nrow, nlay), order="C", dtype=self.dtype)
             vals = np.ma.array(vals, mask=currentmask)
@@ -741,10 +742,10 @@ class GridProperty(_Grid3D):
 
         # the self._isdiscrete property shall win over numpy dtype
         if "int" in str(values.dtype) and not self._isdiscrete:
-            values = values.astype(np.float64)
+            values = np.ma.array(values.astype(np.float64), copy=False)
 
         if "float" in str(values.dtype) and self._isdiscrete:
-            values = values.astype(np.int32)
+            values = np.ma.array(values.astype(np.int32), copy=False)
 
         return values
 
@@ -842,7 +843,7 @@ class GridProperty(_Grid3D):
         propertyname: str,
         realisation: int = 0,
         casting: (
-            Literal["no", "equiv", "safe", "same_kind", "unsafe"] | None
+            Literal["no", "equiv", "safe", "same_kind", "same_value", "unsafe"] | None
         ) = "unsafe",
     ) -> None:
         """
@@ -1014,7 +1015,7 @@ class GridProperty(_Grid3D):
 
         return act
 
-    def get_active_npvalues1d(self) -> np.ma.MaskedArray:
+    def get_active_npvalues1d(self) -> np.ndarray:
         """
         Get the active cells as a 1D numpy masked array.
 
@@ -1027,9 +1028,9 @@ class GridProperty(_Grid3D):
     def get_npvalues1d(
         self,
         activeonly: bool = False,
-        fill_value: npt.ArrayLike = np.nan,
+        fill_value: float | int | str | bytes | complex | np.generic | None = np.nan,
         order: Literal["C", "F"] = "C",
-    ) -> np.ma.MaskedArray:
+    ) -> np.ndarray:
         """
         Return the grid property as a 1D numpy array (copy) for active or all
         cells, but inactive have a fill value.
@@ -1047,10 +1048,12 @@ class GridProperty(_Grid3D):
         .. versionchanged:: 2.8 Added `fill_value` and `order`
 
         """
-        vact = self.values1d.copy()
+        vact: np.ma.MaskedArray = np.ma.array(self.values1d, copy=True)
 
         if order == "F":
-            vact = _gridprop_lowlevel.c2f_order(self, vact)
+            data_f = _gridprop_lowlevel.c2f_order(self, vact.data)
+            mask_f = _gridprop_lowlevel.c2f_order(self, np.ma.getmaskarray(vact))
+            vact = np.ma.MaskedArray(data_f, mask=mask_f, copy=False)
 
         if activeonly:
             return vact.compressed()  # safer than vact[~vact.mask] if no masked

